From c9bbdfd8d1a7d4ef3d251ce3b57e6bcf86589d0d Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Wed, 29 May 2019 10:12:52 +0200
Subject: [PATCH 265/386] telephony: mvebu_phone: tdmmc: enable internal
 loopback configuration

This patch adds an extra flag to the TAL API, with which the
user configure the internal loopback on TX/RX lines.
It can be used for the TDM validation without need
to provide any external connectivity and devices.

Change-Id: I43ab6e95c9898952e7903137d34401f1c64aaf2b
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/10173
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone.h    |  1 +
 drivers/telephony/mvebu_phone/tal/tal.h     |  1 +
 drivers/telephony/mvebu_phone/tal/tal_dev.c |  2 ++
 drivers/telephony/mvebu_phone/tal/tal_dev.h |  1 +
 drivers/telephony/mvebu_phone/tdmmc/tdmmc.c | 17 +++++++++++++----
 drivers/telephony/mvebu_phone/tdmmc/tdmmc.h |  2 ++
 6 files changed, 20 insertions(+), 4 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone.h b/drivers/telephony/mvebu_phone/mv_phone.h
index a93e25b35b06..1609b8236155 100644
--- a/drivers/telephony/mvebu_phone/mv_phone.h
+++ b/drivers/telephony/mvebu_phone/mv_phone.h
@@ -114,6 +114,7 @@ struct mv_phone_params {
 	u16 pcm_slot[32];
 	u8 sampling_period;
 	u16 total_channels;
+	bool enable_internal_loopback;
 };
 
 struct mv_phone_dev {
diff --git a/drivers/telephony/mvebu_phone/tal/tal.h b/drivers/telephony/mvebu_phone/tal/tal.h
index 82edbab7c36d..fa11081d235b 100644
--- a/drivers/telephony/mvebu_phone/tal/tal.h
+++ b/drivers/telephony/mvebu_phone/tal/tal.h
@@ -34,6 +34,7 @@ struct tal_params {
 	u16 pcm_slot[TAL_MAX_PHONE_LINES];
 	u8 sampling_period;
 	u16 total_lines;
+	bool enable_internal_loopback;
 };
 
 struct tal_stats {
diff --git a/drivers/telephony/mvebu_phone/tal/tal_dev.c b/drivers/telephony/mvebu_phone/tal/tal_dev.c
index a0fdb1ec3ffa..c9c1c5736fcb 100644
--- a/drivers/telephony/mvebu_phone/tal/tal_dev.c
+++ b/drivers/telephony/mvebu_phone/tal/tal_dev.c
@@ -171,6 +171,8 @@ static long tal_dev_ioctl(struct file *file_p, unsigned int cmd,
 		tal_params.pcm_format = tal_dev_params.pcm_format;
 		tal_params.sampling_period = 10; /* ms */
 		tal_params.total_lines = tal_dev_params.total_lines;
+		tal_params.enable_internal_loopback =
+					tal_dev_params.enable_internal_loopback;
 		for (i = 0; i < TAL_MAX_PHONE_LINES; i++)
 			tal_params.pcm_slot[i] =
 					    (i + 1) * tal_dev_params.pcm_format;
diff --git a/drivers/telephony/mvebu_phone/tal/tal_dev.h b/drivers/telephony/mvebu_phone/tal/tal_dev.h
index 19ec88f49326..1bc03b8aa28c 100644
--- a/drivers/telephony/mvebu_phone/tal/tal_dev.h
+++ b/drivers/telephony/mvebu_phone/tal/tal_dev.h
@@ -19,6 +19,7 @@
 struct tal_dev_params {
 	unsigned char pcm_format;
 	unsigned short total_lines;
+	unsigned char enable_internal_loopback;
 };
 
 #define	TDM_DEV_TDM_TEST_MODE_ENABLE	_IO(TAL_DEV_IOCTL_MAGIC, 8)
diff --git a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
index 04a904936435..845b964712ff 100644
--- a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
+++ b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
@@ -325,7 +325,7 @@ int tdmmc_init(void __iomem *base, struct device *dev,
 {
 	struct tdmmc_dram_entry *act_dpram_entry;
 	u32 buff_size, chan, total_rx_desc_size, total_tx_desc_size;
-	u32 max_poll, clk_sync_ctrl_reg, count;
+	u32 max_poll, clk_sync_ctrl_reg, count, reg_val;
 	u16 pcm_slot, index;
 	int ret;
 
@@ -639,9 +639,18 @@ int tdmmc_init(void __iomem *base, struct device *dev,
 	writel(clk_sync_ctrl_reg, tdmmc->regs + TDM_CLK_AND_SYNC_CONTROL_REG);
 
 	/* Set TDM TCR register */
-	writel((readl(tdmmc->regs + FLEX_TDM_CONFIG_REG) |
-	       CONFIG_FLEX_TDM_CONFIG),
-	       tdmmc->regs + FLEX_TDM_CONFIG_REG);
+	reg_val = readl(tdmmc->regs + FLEX_TDM_CONFIG_REG) |
+			CONFIG_FLEX_TDM_CONFIG;
+
+	if (tdm_params->enable_internal_loopback) {
+		dev_info(tdmmc->dev,
+			 "enable internal loopback on TX/RX lines\n");
+		reg_val |= TDM_TDIAG_MASK;
+	} else {
+		reg_val &= ~TDM_TDIAG_MASK;
+	}
+
+	writel(reg_val, tdmmc->regs + FLEX_TDM_CONFIG_REG);
 
 	/**********************************************************************/
 	/* Time Division Multiplexing(TDM) Interrupt Controller Configuration */
diff --git a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
index 2af6d92beb78..baf07a029a27 100644
--- a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
+++ b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.h
@@ -169,6 +169,8 @@
 #define TSD_NO_DELAY			(0 << TSD_OFFS)
 #define RSD_OFFS			27
 #define RSD_NO_DELAY			(0 << RSD_OFFS)
+#define TDM_TDIAG_OFFS			29
+#define TDM_TDIAG_MASK			(3 << TDM_TDIAG_OFFS)
 #define TDM_TEN_OFFS			31
 #define TDM_TEN_MASK			(1 << TDM_TEN_OFFS)
 
-- 
2.17.1

